packages/rosmsg-deser/src/DeserializerFactory.ts

```diff
@@ -0,0 +1,319 @@
+          }`;
+      } else {
+        return `
+          // ${field.type}[] ${field.name}
+          get ${field.name}() {
+            const offset = this.${field.name}_offset(this.#view, this.#offset);
+            return readers.dynamicArray(this.#view, offset, ${readerFn}, ${sizeFn});
+          }`;
+      }
+    }
+  } else {
+    return `get ${field.name}() {
+        const offset = this.${field.name}_offset(this.#view, this.#offset);
+        return ${readerFn}(this.#view, offset);
+      }`;
+  }
+}
+
+// Create a SerializedMessageReader
+//
+// The output is a set of classes - one for each custom message type. Only the root message
+// class is exposed.
+//
+// Each LazyMessage class consists of static _size_ functions, _offset_ methods, and property _getters.
+// The size functions calculate the size of fields within arrays.
+// The offset methods calculate the start byte of the field within the entire message buffer.
+// The getter de-serializes the field from the message buffer.
+export default function buildClass(types: RosMsgDefinition[]): SerializedMessageReader {
+  const classes = new Array<string>();
+
+  // Build the types in reverse order so the root message appears last
+  // Since the root message depends on custom types we want those to be defined
+  for (const type of types.reverse()) {
+    const name = sanitizeName(type.name ?? "__RootMsg");
+
+    const offsetFns = new Array<string>();
+    const fields = new Array<string>();
+
+    // getters need to "look back" at the previous field to create the offset function calls
+    let prevField: RosMsgField | undefined;
+
+    for (const field of type.definitions) {
+      // constants have no impact on deserialization
+      if (field.isConstant === true) {
+        continue;
+      }
+
+      // offsets tell you where the raw data of your field starts (including any length bytes)
+      // they are the size of the offset of the previous field + size of previous field
+      // the first first field is at offset 0
+      if (!prevField) {
+        offsetFns.push(`
+          ${field.name}_offset(view, initOffset) {
+            return initOffset;
+          }`);
+      } else {
+        // offsets tell you where the raw data of your field starts (including any length bytes)
+        // they are the size of the offset of the previous field + size of previous field
+        offsetFns.push(`
+          ${field.name}_offset(view, initOffset) {
+            const prevOffset = this.${prevField.name}_offset(view, initOffset);
+            const totalOffset = prevOffset + ${name}.${prevField.name}_size(view, prevOffset);
+  
+            // future calls will avoid re-calculating the offset
+            this.${field.name}_offset = () => totalOffset;
+            return totalOffset;
+          }`);
+      }
+
+      fields.push(`${field.name}: this.${field.name}`);
+
+      prevField = field;
+    }
+
+    const messageSrc = `class ${name} {
```

Comment:
nit: for readability & consistency with some other style
```suggestion
    const messageSrc = `
    class ${name} {
```